Utilities/rANS/src/SymbolStatistics.cxx

```diff
@@ -39,60 +39,48 @@ void SymbolStatistics::rescaleToNBits(size_t bits)
+  std::vector<uint32_t> sortIdx;
+  sortIdx.reserve(mNUsedAlphabetSymbols);
+
   // resample distribution based on cumulative frequencies_
-  for (size_t i = 1; i <= mFrequencyTable.size(); i++)
-    mCumulativeFrequencyTable[i] =
-      (static_cast<uint64_t>(newCumulatedFrequency) *
-       mCumulativeFrequencyTable[i]) /
-      cumulatedFrequencies;
-
-  // if we nuked any non-0 frequency symbol to 0, we need to steal
-  // the range to make the frequency nonzero from elsewhere.
-  //
-  // this is not at all optimal, i'm just doing the first thing that comes to
-  // mind.
   for (size_t i = 0; i < mFrequencyTable.size(); i++) {
-    if (mFrequencyTable[i] &&
-        mCumulativeFrequencyTable[i + 1] == mCumulativeFrequencyTable[i]) {
-      // symbol i was set to zero freq
-
-      // find best symbol to steal frequency from (try to steal from low-freq
-      // ones)
-      std::pair<size_t, size_t> stealFromEntry{mFrequencyTable.size(), ~0u};
-      for (size_t j = 0; j < mFrequencyTable.size(); j++) {
-        uint32_t frequency =
-          mCumulativeFrequencyTable[j + 1] - mCumulativeFrequencyTable[j];
-        if (frequency > 1 && frequency < stealFromEntry.second) {
-          stealFromEntry.second = frequency;
-          stealFromEntry.first = j;
-        }
-      }
-      assert(stealFromEntry.first != mFrequencyTable.size());
-
-      // and steal from it!
-      if (stealFromEntry.first < i) {
-        for (size_t j = stealFromEntry.first + 1; j <= i; j++)
-          mCumulativeFrequencyTable[j]--;
-      } else {
-        assert(stealFromEntry.first > i);
-        for (size_t j = i + 1; j <= stealFromEntry.first; j++)
-          mCumulativeFrequencyTable[j]++;
-      }
+    if (mFrequencyTable[i]) {
+      sortIdx.push_back(i); // we will sort only those memorize only those entries which can be used
     }
   }
 
+  std::sort(sortIdx.begin(), sortIdx.end(), [this](uint32_t i, uint32_t j) { return this->getCumulativeFrequency(i) < this->getCumulativeFrequency(j); });
+  size_t need_shift = 0;
+  for (size_t i = 0; i < sortIdx.size(); i++) {
+    if (static_cast<uint64_t>(getCumulativeFrequency(sortIdx[i])) * (newCumulatedFrequency - need_shift) / cumulatedFrequencies >= 1) {
+      break;
+    }
+    need_shift++;
+  }
+
+  size_t shift = 0;
+  for (size_t i = 0; i < mFrequencyTable.size(); i++) {
+    if (static_cast<uint64_t>(getCumulativeFrequency(i)) * (newCumulatedFrequency - need_shift) / cumulatedFrequencies < 1) {
+      mCumulativeFrequencyTable[i + 1] = mCumulativeFrequencyTable[i] + 1;
+      shift++;
+    } else {
+      mCumulativeFrequencyTable[i + 1] = (static_cast<uint64_t>(newCumulatedFrequency - need_shift) * mCumulativeFrequencyTable[i + 1]) / cumulatedFrequencies + shift;
+    }
+  }
+  assert(shift == need_shift);
+
   // calculate updated freqs and make sure we didn't screw anything up
   assert(mCumulativeFrequencyTable.front() == 0 &&
          mCumulativeFrequencyTable.back() == newCumulatedFrequency);
+
```

Comment:
I think the code should be something like below, but, although now the assert on ``mCumulativeFrequencyTable[i + 1]>mCumulativeFrequencyTable[i]`` is OK, the assert on ``need_shift==shift`` fails miserably (sometimes the difference is huge). 
````
  std::vector<uint32_t> sortIdx;
  sortIdx.reserve(mNUsedAlphabetSymbols);
  
  // resample distribution based on cumulative frequencies_
  for (size_t i = 0; i < mFrequencyTable.size(); i++) {
    if (mFrequencyTable[i]) {
      sortIdx.push_back(i); // we will sort only those memorize only those entries which can be used
    }
  }
  
  std::sort(sortIdx.begin(), sortIdx.end(), [this](uint32_t i, uint32_t j) { return this->getCumulativeFrequency(i) < this->getCumulativeFrequency(j); });
  size_t need_shift = 0;
  for (size_t i = 0; i < sortIdx.size(); i++) {
    uint32_t vA = static_cast<uint64_t>(mCumulativeFrequencyTable[sortIdx[0]]) * (newCumulatedFrequency-need_shift)  / cumulatedFrequencies;
    uint32_t vB = static_cast<uint64_t>(mCumulativeFrequencyTable[sortIdx[i]+1]) * (newCumulatedFrequency-need_shift)  / cumulatedFrequencies;
    if ( vA>vB ) {
      break;
    }
    need_shift++;
  }
  newCumulatedFrequency -= need_shift;
  size_t shift = 0;
  // note that since mCumulativeFrequencyTable[0]==0, the rescaled also will be 0
  for (size_t i = 1; i <= mFrequencyTable.size(); i++) {
    mCumulativeFrequencyTable[i] = (static_cast<uint64_t>(newCumulatedFrequency) * mCumulativeFrequencyTable[i]) / cumulatedFrequencies + shift;
    if (mCumulativeFrequencyTable[i] == mCumulativeFrequencyTable[i-1]) {
      if (!mFrequencyTable[i-1]) { // original freq. was 0
        continue;
      }
      mCumulativeFrequencyTable[i]++;
      shift++;
    }
  }
  assert(shift == need_shift); // RS: this fails!!!! 
  // calculate updated freqs and make sure we didn't screw anything up
  assert(mCumulativeFrequencyTable.front() == 0 && mCumulativeFrequencyTable.back() == newCumulatedFrequency+shift); // RS: this fails!!!! 

  for (size_t i = 0; i < mFrequencyTable.size(); i++) {
    if (mFrequencyTable[i] == 0)
      assert(mCumulativeFrequencyTable[i + 1] == mCumulativeFrequencyTable[i]);
    else
      assert(mCumulativeFrequencyTable[i + 1] > mCumulativeFrequencyTable[i]);
    
    // calc updated freq
    mFrequencyTable[i] = getCumulativeFrequency(i);
  }
````

Reply:
thx for having a look and sorry for not testing properly... will work on this now. Was also tired yesterday.

What concerns commenting out all regularizations: Isn't the support for symbols with 0 probability already there. I think it was merged in the last PR, not #4044. So perhaps this is already just working transparently? So perhaps we are just fine commenting out everything.

Reply:
I think support for 0 prob. symbols can come only with https://github.com/AliceO2Group/AliceO2/pull/4044 (i.e. writing them as "rares" not covered by dictionary), @MichaelLettrich can confirm.